torch/distributed/_composable/_ddp.py

```diff
@@ -179,368 +92,15 @@ def backward(ctx, *grad_outputs):
-                      gradients and ``allreduce`` communication buckets. When
-                      gradients are views, ``detach_()`` cannot be called on the
-                      gradients. If hitting such errors, please fix it by
-                      referring to the :meth:`~torch.optim.Optimizer.zero_grad`
-                      function in ``torch/optim/optimizer.py`` as a solution.
-                      Note that gradients will be views after first iteration, so
-                      the peak memory saving should be checked after first iteration.
-        static_graph (bool): When set to ``True``, DDP knows the trained graph is
-                     static. Static graph means 1) The set of used and unused
-                     parameters will not change during the whole training loop; in
-                     this case, it does not matter whether users set
-                     ``find_unused_parameters = True`` or not. 2) How the graph is trained
-                     will not change during the whole training loop (meaning there is
-                     no control flow depending on iterations).
-                     When static_graph is set to be ``True``, DDP will support cases that
-                     can not be supported in the past:
-                     1) Reentrant backwards.
-                     2) Activation checkpointing multiple times.
-                     3) Activation checkpointing when model has unused parameters.
-                     4) There are model parameters that are outside of forward function.
-                     5) Potentially improve performance when there are unused parameters,
-                     as DDP will not search graph in each iteration to detect unused
-                     parameters when static_graph is set to be ``True``.
-                     To check whether you can set static_graph to be ``True``, one way is to
-                     check ddp logging data at the end of your previous model training,
-                     if ``ddp_logging_data.get("can_set_static_graph") == True``, mostly you
-                     can set ``static_graph = True`` as well.
-
-                     Example::
-                         >>> # xdoctest: +SKIP("undefined variables")
-                         >>> model_DDP = torch.nn.parallel.DistributedDataParallel(model)
-                         >>> # Training loop
-                         >>> ...
-                         >>> ddp_logging_data = model_DDP._get_ddp_logging_data()
-                         >>> static_graph = ddp_logging_data.get("can_set_static_graph")
-
-
-    Attributes:
-        module (Module): the module to be parallelized.
-
-    Example::
-
-        >>> # xdoctest: +SKIP("undefined variables")
-        >>> torch.distributed.init_process_group(backend='nccl', world_size=4, init_method='...')
-        >>> net = torch.nn.parallel.DistributedDataParallel(model)
-    """
-
+class DistributedDataParallel(Module):
```

Comment:
I saw the `replicate` API calls into this module wrapper. Is this necessary? Do you plan to extract states and method from this file and remove the model wrapper later?

```
        self._ddp = _ddp.DistributedDataParallel(
            self._param_list, **self.kwargs
        )
```

Reply:
Yes, will do!